### ISSUE TITLE
[rush] Design: Customize Concurrency By Project (and support device assignment)

### ISSUE THREAD

**Issue by elliot-nelson** [microsoft/rushstack#4456]
## Summary

In a Rush monorepo with a variety of different projects, I would like to be able to customize concurrency by project. For example, I'd like to turn on parallelism for a phased command, but _this particular group of 10 projects_ uses a shared resource during tests, and so I need them to barred from running in parallel _with each other_.

In a related requirement, I may have projects that consume a resource out of a limited pool, such as "test device" (i.e. a simulator target or physical hardware device). If I have a pool of 3 devices, and 30 projects to test on them, then I need to not only set a concurrency limit of 3 for those 30 projects, but also make sure I am "assigning" a device to each project to test on.

In either case, other projects not in a limited pool should be able to keep churning away while Rush waits for limited resources.

## Details

This **WIP** design is light on details. But we can lay out some possibilities.

 - In a GitHub Actions workflow, they accomplish "customized concurrency" via the `concurrency: ` setting. Whatever _value_ this setting returns is considered a lock, essentially. So for example, setting `concurrency: abcd` (a static value) forces all jobs to run sequentially, because they all have the same value. Whereas, setting `concurrency: ${{ matrix.xcode-version }}` in the context of a matrix would allow multiple jobs to run in parallel as long as they had _different_ "xcode-version" values.

 - We need a way to set such a value for a project -- and more usefully, a range of projects. In particular we would want to support using tags to identify projects in various limited resource groups. Ideally, we could configure using JSON selector expressions linked to concurrency values.

 - The topic of "assigning devices" (or some other limited resource like license keys, browser PIDs, remote IPs) is thorny and very company-specific. Passing them to the test runner is also thorny and company-specific, for example, what if there's a pool of 12 devices and we want to select 3 and pass them to each runner? What if we need logic to determine if a device is still alive and ready to receive tests? What if we need a way to fail the tests if no devices are available after X minutes? It seems like we need to offload _as much of this logic as possible_ to the monorepo maintainer.

 - If this feature is so advanced that we can't support it just with config files and script hooks, then I think requiring a Rush plugin is reasonable -- the question is how to integrate hooks for such a plugin into the task scheduler so that the maintainer can write one.

Thanks for considering!

### Design 1: "Special Exit Code"

In an early discussion about this feature years ago, @octogonz proposed that we could give Rush a tiny change: you could configure a special exit code (for example, exit code 37) to mean not "I have failed" but rather "I am not ready yet".

With this feature, you could push _all_ the logic described above into, say, a custom node script wrapper in your project's `_phase:test` script. It's the job of your script to go decide, somehow, whether it's safe to run, and if it is, to run. If it's _not_ safe to run, you would exit with exit code 37, and Rush would bump you to the back of its potential runnable tasks, and assign some value (e.g. 10 seconds) before it would attempt to run you again.

PROS:

 - Elegant and minimal from Rush's perspective, no huge design discussions necessary.
 - Relatively easy API -- no "plugins" to build, etc.

CONS:

 - How to lock/mutex etc. totally left to user, potentially leading to badly designed test scripts
 - Keeping track of which devices have been assigned, etc. is all up to individual runs of the test scripts to manage, requiring the monorepo maintainer to build some kind of external lock, rely on a series of file system lockfiles, etc.
 - No good way to customize how many times Rush will try or how long it will wait between tries.
 - Once Rush runs out of tasks that are ready, it is just in a loop of starting up a bunch of tasks that return exit code 37 and waiting for one of them to run... what does the console log output look like in these scenarios?

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| Would you consider contributing a PR? | Yes |

**Comment by octogonz** [comment]
Could you add some concrete examples of how this feature might be used, along with hypothetical config snippets?  Rush managing everything sounds great in theory, but the problem of lab resources probably has lots of little details that will be different for every scenario, so I question whether it is really possible to do better than a bunch of custom scripts for each setup.

**Comment by elliot-nelson** [comment]
@octogonz Good idea, I have added a Use Cases section with my two use cases, and tried to make them as concrete as possible.

**Comment by dmichon-msft** [comment]
The Rush Cobuilds feature introduced the same concept of "try to get a lock, and if you fail, defer the Operation", so that logic would be something we should consolidate and generalize. I could easily see as a stress test edge case a build graph where each individual Operation is in several separate concurrency buckets (at a minimum there will be 2: the global one and the specific one).

This is a good opportunity for Rush to converge onto `@rushstack/operation-graph`, since this requires scheduler rework, and that will allow Heft to benefit from the same logic. Since the work queue in `@rushstack/operation-graph` already supports asynchronous scheduling, it is less complex to extend with resource locking.
